pgjdbc/src/main/java/org/postgresql/jdbc/TimestampUtils.java

```diff
@@ -633,8 +634,18 @@ private static void appendTime(StringBuilder sb, int hours, int minutes, int sec
     }
     sb.append('.');
     int len = sb.length();
-    sb.append(nanos / 1000); // append microseconds
-    int needZeros = 6 - (sb.length() - len);
+    int micros = nanos / 1000;
+    int microsDigits = 0;
+    for (int i = 0; i < 6; i++) {
+      if (micros % POWERS10[i + 1] != 0) {
```

Comment:
Frankly speaking, I would just append nanos as is, then walk backwards and remove zeros one by one.